File: flash_mastery_api/src/main/java/com/flash/mastery/entity/StudySession.java
package com.flash.mastery.entity;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import com.flash.mastery.constant.NumberConstants;
import com.flash.mastery.entity.enums.StudyMode;
import com.flash.mastery.entity.enums.StudySessionStatus;

import jakarta.persistence.CascadeType;
import jakarta.persistence.CollectionTable;
import jakarta.persistence.Column;
import jakarta.persistence.ElementCollection;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Entity
@Table(name = "study_sessions")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class StudySession extends BaseAuditEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "deck_id", nullable = false)
    @ToString.Exclude
    private Deck deck;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "study_session_flashcard_ids", joinColumns = @JoinColumn(name = "session_id"))
    @Column(name = "flashcard_id")
    @Default
    private List<UUID> flashcardIds = new ArrayList<>();

    @Enumerated(EnumType.STRING)
    @Column(name = "current_mode", nullable = false)
    @Default
    private StudyMode currentMode = StudyMode.OVERVIEW;

    @Column(name = "current_batch_index", nullable = false)
    @Default
    private Integer currentBatchIndex = 0;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    @Default
    private StudySessionStatus status = StudySessionStatus.IN_PROGRESS;

    @OneToMany(mappedBy = "session", cascade = CascadeType.ALL, orphanRemoval = true)
    @Default
    @ToString.Exclude
    private List<StudySessionProgress> progressRecords = new ArrayList<>();

    @Column(name = "completed_at")
    private java.time.LocalDateTime completedAt;

    /**
     * Ensure status is never null before persisting or updating.
     */
    @PrePersist
    @PreUpdate
    private void ensureStatus() {
        if (this.status == null) {
            this.status = StudySessionStatus.IN_PROGRESS;
        }
    }

    /**
     * Get flashcards for current batch.
     */
    public List<UUID> getCurrentBatch() {
        final var batchSize = NumberConstants.STUDY_SESSION_BATCH_SIZE;
        final var start = this.currentBatchIndex * batchSize;
        final var end = Math.min(start + batchSize, this.flashcardIds.size());
        if (start >= this.flashcardIds.size()) {
            return new ArrayList<>();
        }
        return this.flashcardIds.subList(start, end);
    }

    /**
     * Check if current batch is complete.
     */
    public boolean isCurrentBatchComplete() {
        final var batch = getCurrentBatch();
        return batch.stream()
                .allMatch(id -> this.progressRecords.stream()
                        .filter(p -> p.getFlashcard().getId().equals(id))
                        .findFirst()
                        .map(p -> p.isModeCompleted(this.currentMode))
                        .orElse(false));
    }

    /**
     * Get next mode in sequence.
     */
    public StudyMode getNextMode() {
        return switch (this.currentMode) {
        case OVERVIEW -> StudyMode.MATCHING;
        case MATCHING -> StudyMode.GUESS;
        case GUESS -> StudyMode.RECALL;
        case RECALL -> StudyMode.FILL_IN_BLANK;
        case FILL_IN_BLANK -> null; // All modes completed
        default -> null;
        };
    }
}

File: lib/domain/entities/study_session.dart
import 'package:equatable/equatable.dart';
import 'package:flash_mastery/domain/entities/study_mode.dart';

/// Study session entity representing a study session.
class StudySession extends Equatable {
  final String id;
  final String deckId;
  final List<String> flashcardIds;
  final StudyMode currentMode;
  final int currentBatchIndex;
  final DateTime startedAt;
  final DateTime? completedAt;
  final Map<String, StudyProgress> progress;

  const StudySession({
    required this.id,
    required this.deckId,
    required this.flashcardIds,
    required this.currentMode,
    this.currentBatchIndex = 0,
    required this.startedAt,
    this.completedAt,
    this.progress = const {},
  });

  StudySession copyWith({
    String? id,
    String? deckId,
    List<String>? flashcardIds,
    StudyMode? currentMode,
    int? currentBatchIndex,
    DateTime? startedAt,
    DateTime? completedAt,
    Map<String, StudyProgress>? progress,
  }) {
    return StudySession(
      id: id ?? this.id,
      deckId: deckId ?? this.deckId,
      flashcardIds: flashcardIds ?? this.flashcardIds,
      currentMode: currentMode ?? this.currentMode,
      currentBatchIndex: currentBatchIndex ?? this.currentBatchIndex,
      startedAt: startedAt ?? this.startedAt,
      completedAt: completedAt ?? this.completedAt,
      progress: progress ?? this.progress,
    );
  }

  /// Get flashcards for current batch (max 7 cards).
  List<String> getCurrentBatch() {
    const batchSize = 7;
    final start = currentBatchIndex * batchSize;
    final end = (start + batchSize).clamp(0, flashcardIds.length);
    return flashcardIds.sublist(start, end);
  }

  /// Check if current batch is complete.
  bool isCurrentBatchComplete() {
    final batch = getCurrentBatch();
    return batch.every((id) => progress[id]?.isCompleted ?? false);
  }

  /// Check if all flashcards are studied.
  bool isComplete() {
    return flashcardIds.every((id) => progress[id]?.isCompleted ?? false);
  }

  /// Get next mode in sequence.
  StudyMode? getNextMode() {
    switch (currentMode) {
      case StudyMode.overview:
        return StudyMode.matching;
      case StudyMode.matching:
        return StudyMode.guess;
      case StudyMode.guess:
        return StudyMode.recall;
      case StudyMode.recall:
        return StudyMode.fillInBlank;
      case StudyMode.fillInBlank:
        return null; // All modes completed
    }
  }

  @override
  List<Object?> get props => [
        id,
        deckId,
        flashcardIds,
        currentMode,
        currentBatchIndex,
        startedAt,
        completedAt,
        progress,
      ];
}

/// Study progress for a single flashcard.
class StudyProgress extends Equatable {
  final String flashcardId;
  final Map<StudyMode, bool> modeCompletion;
  final int correctAnswers;
  final int totalAttempts;
  final DateTime? lastStudiedAt;

  const StudyProgress({
    required this.flashcardId,
    this.modeCompletion = const {},
    this.correctAnswers = 0,
    this.totalAttempts = 0,
    this.lastStudiedAt,
  });

  StudyProgress copyWith({
    String? flashcardId,
    Map<StudyMode, bool>? modeCompletion,
    int? correctAnswers,
    int? totalAttempts,
    DateTime? lastStudiedAt,
  }) {
    return StudyProgress(
      flashcardId: flashcardId ?? this.flashcardId,
      modeCompletion: modeCompletion ?? this.modeCompletion,
      correctAnswers: correctAnswers ?? this.correctAnswers,
      totalAttempts: totalAttempts ?? this.totalAttempts,
      lastStudiedAt: lastStudiedAt ?? this.lastStudiedAt,
    );
  }

  /// Check if flashcard is completed in all modes.
  bool get isCompleted {
    return modeCompletion.length == StudyMode.values.length &&
        modeCompletion.values.every((completed) => completed);
  }

  /// Mark a mode as completed.
  StudyProgress markModeCompleted(StudyMode mode) {
    final updated = Map<StudyMode, bool>.from(modeCompletion);
    updated[mode] = true;
    return copyWith(
      modeCompletion: updated,
      lastStudiedAt: DateTime.now(),
    );
  }

  /// Record an answer attempt.
  StudyProgress recordAttempt({required bool isCorrect}) {
    return copyWith(
      correctAnswers: isCorrect ? correctAnswers + 1 : correctAnswers,
      totalAttempts: totalAttempts + 1,
      lastStudiedAt: DateTime.now(),
    );
  }

  @override
  List<Object?> get props => [
        flashcardId,
        modeCompletion,
        correctAnswers,
        totalAttempts,
        lastStudiedAt,
      ];
}

